xonsh/events.py

```diff
@@ -0,0 +1,113 @@
+"""
+Events for xonsh.
+
+In all likelihood, you want builtins.__xonsh_events__
+
+The best way to "declare" an event is something like::
+
+    __xonsh_events__.on_spam.__doc__ = "Comes with eggs"
+"""
+
+
+class Event(set):
+    """
+    A given event that handlers can register against.
+
+    Acts as a ``set`` for registered handlers.
+
+    Note that ordering is never guarenteed.
+    """
+    def __init__(self, doc=None):
+        self.__doc__ = doc
+
+    def handler(self, callable):
+        """
+        Registers a handler. It's suggested to use this as a decorator.
+        """
+        self.add(callable)
+        return callable
+
+    def calleach(self, *pargs, **kwargs):
+        """
+        The core handler caller that all others build on.
+
+        This works as a generator. Each handler is called in turn and its
+        results are yielded.
+
+        If the generator is interupted, no further handlers are called.
+
+        The caller may send() new positional arguments (eg, to implement
+        modifying semantics). Keyword arguments cannot be modified this way.
+        """
+        for handler in self:
+            newargs = yield handler(*pargs, **kwargs)
```

Comment:
On the whole "validators" think I think that's not pythonic let's keep it
simpler.  Someone said (and they're right) event handlers can exit on line
one if they've got a good validation function that applies.  Same
difference.

Xonsh is about providing outer infrastructure for future python+shell
end-user needs.

Letting imported modules do something new and better in xonsh than they do
in python is an exciting idea.

Maybe just like modules have __all__ for all exported functions in python,
in xonsh they could have __events__ for exported event handlers?

Kind regards,


*Paul Reiber*http://about.me/reiber

On Fri, Jul 29, 2016 at 4:33 PM, Anthony Scopatz <notifications@github.com>
wrote:

> In xonsh/events.py
> <https://github.com/xonsh/xonsh/pull/1501#discussion_r72874787>:
>
> > +        return callable
> > +
> > +    def calleach(self, *pargs, **kwargs):
> > +        """
> > +        The core handler caller that all others build on.
> > +
> > +        This works as a generator. Each handler is called in turn and its
> > +        results are yielded.
> > +
> > +        If the generator is interupted, no further handlers are called.
> > +
> > +        The caller may send() new positional arguments (eg, to implement
> > +        modifying semantics). Keyword arguments cannot be modified this way.
> > +        """
> > +        for handler in self:
> > +            newargs = yield handler(*pargs, **kwargs)
>
> I would call this deterministic but chaotic, and not easily reasoned about.
>
> Exactly. I feel that this needs to be easily reasoned about. It is a
> public interface.
>
> Which makes me think that for the sake of future sanity, any kind of event
> semantics that depend on ordering should be avoided.
>
> Yes, I agree. These kinds of things should be avoided if at all possible.
> However, avoided doesn't mean forbidden. There may even be cases in the
> core that may be most easily implemented in this way. I don't think we need
> to implement a full DAG dependency solver for this, but we also don't want
> to explicitly disallow any ad hoc dependencies either.
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/xonsh/xonsh/pull/1501/files/3bb099b05b27208d755b66c3732497a599d86077#r72874787>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AADlQJIMDJf7H_gnLFtEEL6WXNpfvxJvks5qao3fgaJpZM4JW3ZE>
> .
>